String "Inf" should be consistently converted to IV/UV

t-a-k opened the pull request:
I found that using string `"Inf"` in contexts that requires IV/UV is converted to wrong results for the first time:
```
$ perl -wle 'my $x = "Inf"; print $x << 0; print $x << 0; print $x << 0'
0
18446744073709551615
18446744073709551615
```
To make matters worse, `"Inf"` seems to "leak" the previous value of the variable:
```
$ perl -wle 'my $x = 42; $x = "Inf"; print $x << 0; print $x << 0'
42
18446744073709551615
```
This PR will fix this problem (and add testcases).

t-a-k committed:
sv.c: String "Inf" is now consistently converted to IV/UV

S_sv_2iuv_common() used to fail to set IV/UV slot on direct conversion
from string "Inf" to IV/UV, so such conversion would return garbage on
the memory.  Now it will return the same result with two-staged
conversions, string to NV first and then NV to IV/UV.

hvds commented on sv.c:
```diff
@@ -2418,7 +2419,7 @@ S_sv_2iuv_common(pTHX_ SV *const sv)
                 not_a_number(sv);
             S_sv_setnv(aTHX_ sv, numtype);
-            return FALSE;
+            goto got_nv;        /* Fill IV/UV slot and set IOKp */
```
I appreciate this should be safe with current core, but it takes some thinking to validate that and feels like a trap that could cause problems in the future. A viable alternative might be to duplicate the `(void)SvIOKp_on(sv)` and move the label to line 2353 (ie the `else` branch of `if (SvNVX(sv) < (NV)IV_MAX + 0.5)`.

t-a-k replied:
I think that it should not jump into the `else` branch directly, because NV may be `-Inf` here.  Furthermore, doing so also skips the `NAN_COMPARE_BROKEN` workaround.

hvds replied:
Ah you're quite right, I failed to pick up the right bits of context; looking at the function as a whole this makes sense.

t-a-k committed:
Add tests for string "-Inf" to IV conversion.

While a numeric -Inf is treated as IV_MIN in IV context (such as
operands for bitwise operators under "use integer"), string "-Inf"
used to yield some random (zero, or previously assigned) value in
such context.

t-a-k commented:
Sorry to trouble you, but I just added another test to `t/op/infnan.t` for checking conversion of `"-Inf"`.

This test might have to be revised in the future if some change were made as the result of this reconsideration...

> (We may also want to reconsider what `-Inf` should yield in integer context, but that's not the job of this PR.)

hvds commented on t/op/infnan.t:
```diff
@@ -576,4 +576,20 @@ cmp_ok('-1e-9999', '==', 0,     "underflow to 0 (runtime) from neg");
 
+# "-Inf" should be converted to IV consistently
+{
+    # Enclosed in a string eval so that majority of infnan.t are run
+    # before loading integer.pm, just in case.
+    eval <<'EOC';
```
Just in case of what? If there's a risk of `use integer` having buggy effects outside its scope, that's not really the responsibility of this test file to cope with. But I see no reason to believe there is such a risk - and certainly we don't use such caution in other much more fundamental test files.

(Also if there were good reason to keep the eval, you should at the very least check for it dying. But far better to remove it.)

t-a-k replied:
> If there's a risk of `use integer` having buggy effects outside its scope, that's not really the responsibility of this test file to cope with.

That makes sense.  I was too worried about becoming the first to `use integer` in `infnan.t`.  I'll push a commit to remove `eval`.

t-a-k committed:
t/op/infnan.t: Remove enclosing `eval` in "-Inf" to IV conversion test

As per review by @hvds in GH #19404, this test file doesn't have to worry
about whether `use integer` might fail.
